x/nexus/keeper/msg_dispatcher.go

```diff
@@ -55,17 +55,25 @@ func (m Messenger) DispatchMsg(ctx sdk.Context, contractAddr sdk.AccAddress, _ s
 }
 
 func (m Messenger) routeMsg(ctx sdk.Context, msg exported.WasmMessage) error {
+	if err := msg.ValidateBasic(); err != nil {
+		return err
+	}
+
 	destinationChain, ok := m.GetChain(ctx, msg.DestinationChain)
 	if !ok {
 		return fmt.Errorf("recipient chain %s is not a registered chain", msg.DestinationChain)
 	}
 
-	id, _, _ := m.GenerateMessageID(ctx)
+	// If message already exists, then this is a no-op to avoid causing an error from reverting the whole message batch being routed in Amplifier
+	if _, ok := m.Nexus.GetMessage(ctx, msg.ID); ok {
+		return nil
+	}
+
 	sourceChain := exported.Chain{Name: msg.SourceChain, SupportsForeignAssets: false, KeyType: tss.None, Module: wasmtypes.ModuleName}
 	sender := exported.CrossChainAddress{Chain: sourceChain, Address: msg.SourceAddress}
 	recipient := exported.CrossChainAddress{Chain: destinationChain, Address: msg.DestinationAddress}
 
-	nexusMsg := exported.NewGeneralMessage(id, sender, recipient, msg.PayloadHash, msg.SourceTxID, msg.SourceTxIndex, nil)
+	nexusMsg := exported.NewGeneralMessage(fmt.Sprintf("%s-%s", msg.SourceChain, msg.ID), sender, recipient, msg.PayloadHash, msg.SourceTxID, msg.SourceTxIndex, nil)
```

Comment:
How can it be ambiguous? We forbid `_` in message ids in amplifier, so I'd rather not add it here.

Reply:
Hmm. It can be ambiguous since both source chain and message ID is allowed to have `-` in it (e.g. `ethereum-2`). Current types of msg.ID (hex or base64 with index) can only contain one `-`, so it won't be ambiguous, but that could change in the future to allow more than one (though probably unlikely).

While we use `_` to concatenate such strings elsewhere, it doesn't work here since the output is the messageId used by core, which is required to not contain `_`.

I'm fine with using `-` as an alternative, although we should avoid allowing more than one `-` in msg.ID for future msg types

Reply:
yes the GeneralMessage struct will fail validation if we use `_` here.

I think the chances of having a collision here are extremely unlikely, because we only allow certain registered message id formats in amplifier. A chain would need to be using message ids that start with `x-` where `x` is allowed to be something that can collide with the suffix of another registered chain name, and that other registered chain is using message ids that can be a suffix of message ids used by the first chain. I think even just not allowing numbers in chain names would be sufficient. But regardless, any new message id formats will be heavily scrutinized, so I have confidence this collision won't happen.

If you are really concerned, we could just go back to generating the message id, or we could append or prepend an incrementing nonce.